Client/SafeBrowsing/GSafeBrowsingDatabase.swift

```diff
@@ -0,0 +1,509 @@
+            })
+            self.mainContext.reset()
+        } catch {
+            //Remove everything and re-create the database
+            self.destroy()
+            self.persistentContainer = self.create()
+
+            log.error("Safe-Browsing: \(error)")
+        }
+    }
+    
+    private lazy var persistentContainer: NSPersistentContainer = {
+        return self.create()
+    }()
+    
+    private lazy var backgroundContext: NSManagedObjectContext = {
+        let context = persistentContainer.newBackgroundContext()
+        context.mergePolicy = NSMergeByPropertyObjectTrumpMergePolicy
+        return context
+    }()
+    
+    private lazy var mainContext: NSManagedObjectContext = {
+        let context = persistentContainer.viewContext
+        context.mergePolicy = NSMergeByPropertyObjectTrumpMergePolicy
+        context.automaticallyMergesChangesFromParent = true
+        return context
+    }()
+    
+    private func save() {
+        saveContext(mainContext)
+    }
+    
+    private func saveContext(_ context: NSManagedObjectContext) {
+        if context.hasChanges {
+            do {
+                try context.save()
+            } catch {
+                log.error("Error Saving Context: \(error)")
+            }
+        }
+    }
+    
+    func canFind() -> Bool {
+        return Date().timeIntervalSince(self.lastFindDate) >= self.lastFindWaitDuration
+    }
+    
+    func canUpdate() -> Bool {
+        return Date().timeIntervalSince(self.lastFetchDate) >= self.lastFetchWaitDuration
+    }
+    
+    func enterBackoffMode(_ mode: BackoffMode) {
+        if mode == .find {
+            let duration = calculateBackoffTime(self.numberOfFindRetries)
+            self.numberOfFindRetries += 1
+            
+            if duration > 0 {
+                self.lastFindDate = Date()
+                self.lastFindWaitDuration = duration
+                self.updateDatabaseInfo { error in
+                    if let error = error {
+                        log.error(error)
+                    }
+                }
+            }
+        } else if mode == .update {
+            let duration = calculateBackoffTime(self.numberOfFetchRetries)
+            self.numberOfFetchRetries += 1
+            
+            if duration > 0 {
+                self.lastFetchDate = Date()
+                self.lastFetchWaitDuration = duration
+                self.updateDatabaseInfo { error in
+                    if let error = error {
+                        log.error(error)
+                    }
+                }
+            }
+        }
+    }
+    
+    func getState(_ type: ThreatType) -> String {
+        var state = ""
+        let backgroundContext = { () -> NSManagedObjectContext in
+            let context = self.persistentContainer.newBackgroundContext()
+            context.mergePolicy = NSMergeByPropertyObjectTrumpMergePolicy
+            return context
+        }()
+        
+        backgroundContext.performAndWait {
```

Comment:
This gets called on main thread anyway

Reply:
It can be called as a result to `fetch` or `update` from any thread when an update is `scheduled` :( That's why I did it like this and I need it synchronous and non-escaping. Calling `perform` will make everything escaping and require callbacks.